tests/lib/__snapshots__/scope-analysis.js.snap

```diff
@@ -2927,6 +2927,255 @@ Object {
 }
 `;
 
+exports[`TypeScript scope analysis tests/fixtures/scope-analysis/identifier-decorators.ts 1`] = `
+Object {
+  "$id": 7,
+  "block": Object {
+    "range": Array [
+      0,
+      65,
+    ],
+    "type": "Program",
+  },
+  "childScopes": Array [
+    Object {
+      "$id": 6,
+      "block": Object {
+        "range": Array [
+          0,
+          65,
+        ],
+        "type": "Program",
+      },
+      "childScopes": Array [
+        Object {
+          "$id": 5,
+          "block": Object {
+            "range": Array [
+              7,
+              64,
+            ],
+            "type": "ClassDeclaration",
+          },
+          "childScopes": Array [
+            Object {
+              "$id": 4,
+              "block": Object {
+                "range": Array [
+                  35,
+                  62,
+                ],
+                "type": "FunctionExpression",
+              },
+              "childScopes": Array [],
+              "functionExpressionScope": false,
+              "isStrict": true,
+              "references": Array [],
```

Comment:
are you sure about this? `Decorator` is applied on top for property of function not on function itself.

Reply:
ok you are right its wong

Reply:
Looks like issue is related only to constructor and not to functions

Reply:
Decorators syntax uses variables. For example, `@dec` refers the variable `dec`, so it creates a `reference` to `dec`.

Reply:
yes i know that. but i have no clue where is an issue, from visitor point of view it looks ok. 

i did some testing if `Decorator` is triggered by 

```
'ClassDeclaration MethodDefinition FunctionExpression Identifier Decorator'
```
and its look fine

Reply:
Mmm, I will investigate it, too...

Reply:
from what i understand it should be here not in function :>

https://github.com/eslint/typescript-eslint-parser/pull/554/files/3aae6574403ec33b493ebc4a0cb1b5bacd24eee2#diff-9641edf4e2df2a2faf3854f97288c829R3037

Reply:
That is the `references` of the `config` variable. But no one uses the `config` variable, so it should be empty :)

`Scope#references` property contains all references in the scope. If a reference couldn't be resolved in the scope, the reference is contained in `Scope#through` as well, then it tries to resolve the reference in the upper scopes. Finally, the global scope's `Scope#through` contains the references of undefined variables. The expected `Decorator` reference should be so. It's a reference of undefined variable.

Reply:
I found the reason.

The `Identifier`s which makes variables are visited in `PatternVisitor` rather than `Referencer`.
So we have to address the parameter decorators in `PatternVisitor`.
Maybe...

```js
class PatternVisitor extends OriginalPatternVisitor {
    Identifier(node) {
        super.Identifier(node);
        if (node.decorators) {
            this.rightHandNodes.push(...node.decorators);
        }
        if (node.typeAnnotation) {
            this.rightHandNodes.push(node.typeAnnotation);
        }
    }

    ArrayPattern(node) {
        node.elements.forEach(this.visit, this);
        if (node.decorators) {
            this.rightHandNodes.push(...node.decorators);
        }
        if (node.typeAnnotation) {
            this.rightHandNodes.push(node.typeAnnotation);
        }
    }

    ObjectPattern(node) {
        node.properties.forEach(this.visit, this);
        if (node.decorators) {
            this.rightHandNodes.push(...node.decorators);
        }
        if (node.typeAnnotation) {
            this.rightHandNodes.push(node.typeAnnotation);
        }
    }
}
```

Reply:
hmm looks good, it's referencing to variable, 
```
 +               "references": Array [
    +                 Object {
    +                   "$id": 4,
    +                   "from": Object {
    +                     "$ref": 5,
    +                   },
    +                   "identifier": Object {
    +                     "name": "Decorator",
    +                     "range": Array [
    +                       37,
    +                       46,
    +                     ],
    +                     "type": "Identifier",
    +                   },
    +                   "kind": "r",
    +                   "resolved": null,
    +                   "writeExpr": undefined,
    +                 },
    +               ],
    +               "throughReferences": Array [
    +                 Object {
    +                   "$ref": 4,
    +                 },
    +               ],
```

when variable has `"$ref": 5,`

Reply:
thank you for help